arch/arm/core/swap_helper.S

```diff
@@ -182,13 +182,14 @@ _thread_irq_disabled:
     vldmia r0, {s16-s31}
 #endif
 
-#ifdef CONFIG_MPU_STACK_GUARD
-    /* r2 contains k_thread */
-    add r0, r2, #0
-    push {r2, lr}
-    blx configure_mpu_stack_guard
-    pop {r2, lr}
-#endif /* CONFIG_MPU_STACK_GUARD */
+#if defined (CONFIG_ARM_MPU)
+    /* Re-program dynamic memory map */
+    push {r0,r1,r2,r3,lr}
+    ldr r0, =_kernel
+    ldr r0, [r0, #_kernel_offset_to_current]
+    bl _arch_configure_dynamic_mpu_regions
+    pop {r0,r1,r2,r3,lr}
```

Comment:
Not my business to nit over ARM assembly, obviously, but this looks like you're doing too much work.  The earlier code seems to know that it has the thread pointer in r2 already (i.e. it doesn't have to fetch it from kernel._current) and that r0/r1/r3 are already scratch and don't need to be saved.

I'd look carefully to see if you really need to be doing all this; context switching is the hottest of hot paths.